ruby/phase-1.md

```diff
@@ -0,0 +1,434 @@
+end
+```
+
+Notes:
+
+- This configuration will get picked by the singleton client
+- This configuration will be used by the worker (unless explicitly overridden)
+- This is not the full list of available configuration options
+- All the configuration options come with sane defaults
+
+For use-cases that require more flexibility, a configuration can be explicitly created for each
+client and worker:
+
+```ruby
+config_1 = Temporal::Configuration.new(
+  url: 'host-1.com:7233',
+  namespace: 'namespace-1',
+  task_queue: 'queue-1'
+)
+client_1 = Temporal::Client.new(config_1)
+worker_1 = Temporal::Worker.new(config_1)
+
+config_2 = Temporal::Configuration.new(
+  url: 'host-2.com:7233',
+  namespace: 'namespace-2',
+  task_queue: 'queue-2'
+)
+client_2 = Temporal::Client.new(config_2)
+worker_2 = Temporal::Worker.new(config_2)
+```
+
+
+## Activity/Workflow execution attributes
+
+This is an important and novel (as far as all the other SDKs are concerned) concept that allows to
+reduce the explicitness of using the Ruby SDK. This is achieved via a few assumptions that have been
+tested against the real world use-cases, but obviously are up for a debate. These are:
+
+- The simplest __viable__ use-case can be satisfied with a single namespace and single task queue
+- Most workflow/activity names are identical to the names of the classes implementing them
+- The most popular setup will be based on a one-to-one mapping between workflows/activities and a
+  namespace + task queue combination
+- In most cases timeouts and retry policies are defined per workflow/activity and not a specific
+  execution. That is to say that workflow/activity owners will probably have the best idea of which
+  timeouts and retry policies should be used when executing it
+
+With that in mind we're proposing a way to configure the said attribute per workflow/activity
+definition. This would reside in the outer API and would look something like this:
+
+```ruby
+class NewWorkflow < Temporal::Workflow
+  namespace 'ns-1'
+  task_queue 'tq-1'
+  timeouts run: 10 * 60 * 60, execution: 20 * 60 * 60
+
+  def execute
+    ...
+  end
+end
+
+class AnotherActivity < Temporal::Activity
+  namespace 'ns-2'
+  task_queue 'tq-2'
+  retry_policy(
+    interval: 1,
+    backoff: 1,
+    max_attempts: 3,
+    non_retriable_errors: [FatalError]
+  )
+
+  def execute
+    ...
+  end
+end
+```
+
+The main idea here is when these are started/scheduled the caller doesn't have to explicitly specify
+`name`, `namespace`, `task_queue`, `timeouts` or `retry_policy`, but can instead rely on the
+existing definitions. The same exact concept applies to registering these classes with a worker,
+that can then figure out which namespace and task queue need to be polled. Here's what that looks
+like:
+
+```ruby
+# starting a workflow
+client.start_workflow(NewWorkflow) # assumes namespace: 'ns-1', task_queue: 'tq-1'
```

Comment:
I understand that this is a bit different from other SDKs. How do you feel about asking people who are already using the existing Ruby SDK for their opinion on this? I think it would be a failure of this project if people would end up building another gem (and that becoming the de facto standard, I've seen many examples of this in the past) that would implementing all the defaults/approaches that you're trying to remove from this proposal. But I'm happy to respect your vision